src/appAccess.ts

```diff
@@ -0,0 +1,288 @@
+import { Address } from './encoding/address.js';
+import {
+  BoxReference,
+  HoldingReference,
+  LocalsReference,
+  ResourceReference,
+} from './types/transactions/base.js';
+import { ensureSafeUnsignedInteger } from './utils/utils.js';
+
+/**
+ * resourceReferencesToEncodingData translates an array of ResourceReferences into an array of encoding data
+ * maps.
+ */
+export function resourceReferencesToEncodingData(
+  appIndex: bigint,
+  references: ReadonlyArray<ResourceReference>
+): Array<Map<string, unknown>> {
+  const accessList: Array<Map<string, unknown>> = [];
+
+  function ensure(target: ResourceReference): number {
+    for (let idx = 0; idx < accessList.length; idx++) {
+      const a = accessList[idx];
+      if (
+        a.get('d') === target.address &&
+        a.get('s') === target.assetIndex &&
+        a.get('p') === target.appIndex
+      ) {
+        return idx + 1; // 1-based index
+      }
+    }
+    if (target.address) {
+      accessList.push(new Map<string, unknown>([['d', target.address]]));
+    }
+    if (target.assetIndex) {
+      accessList.push(new Map<string, unknown>([['s', target.assetIndex]]));
+    }
+    if (target.appIndex) {
+      accessList.push(new Map<string, unknown>([['p', target.appIndex]]));
+    }
+    return accessList.length; // length is 1-based position of new element
+  }
+
+  const zeroAddr = Address.zeroAddress();
+  for (const rr of references) {
+    if (rr.address || rr.assetIndex || rr.appIndex) {
+      ensure(rr);
+    }
+
+    if (rr.holding) {
```

Comment:
Perhaps we should `continue` at the end of each `if (rr.xxxx) {` block? Or perhaps make these cascading `if else`s?

Reply:
done